include/nbl/builtin/hlsl/utils/surface_transform.hlsl

```diff
@@ -0,0 +1,177 @@
+    }
+}
+
+//! Use this function to apply the swapchain tranformation to the screenspace coordinate `coord` 
+//! Usecase = [Scatter]:
+//!   When directly writing to your swapchain using `imageStore` in order to match the orientation of the device relative to it's natural orientation. 
+//! Warning: Be aware that almost always you'd want to do a single transform in your rendering pipeline.
+int2 applyToScreenSpaceCoordinate(in uint swapchainTransform, in int2 coord, in int2 screenSize)
+{
+    int2 lastTexel = screenSize - (1).xx;
+    switch (swapchainTransform) 
+    {
+    case IDENTITY:
+        return coord;
+    case ROTATE_90:
+        return int2(coord.y, lastTexel.x - coord.x);
+    case ROTATE_180:
+        return int2(lastTexel) - coord;
+    case ROTATE_270:
+        return int2(lastTexel.y - coord.y, coord.x);
+    case HORIZONTAL_MIRROR:
+        return int2(lastTexel.x - coord.x, coord.y);
+    case HORIZONTAL_MIRROR_ROTATE_90:
+        return coord.yx;
+    case HORIZONTAL_MIRROR_ROTATE_180:
+        return int2(coord.x, lastTexel.y - coord.y);
+    case HORIZONTAL_MIRROR_ROTATE_270:
+        return lastTexel - coord.yx;
+    default:
+        return (0).xx;
+    }
+}
+
+//! [width,height] might switch to [height, width] in orientations such as 90°CW
+//! Usecase: Currently none in the shaders
+int2 transformedExtents(in uint swapchainTransform, in int2 screenSize)
+{
+    switch (swapchainTransform) 
+    {
+    case IDENTITY:
+    case HORIZONTAL_MIRROR:
+    case HORIZONTAL_MIRROR_ROTATE_180:
+    case ROTATE_180:
+        return screenSize;
+    case ROTATE_90:
+    case ROTATE_270:
+    case HORIZONTAL_MIRROR_ROTATE_90:
+    case HORIZONTAL_MIRROR_ROTATE_270:
+        return screenSize.yx;
+    default:
+        return (0).xx;
+    }
+}
+
+// TODO: surface_transform::transformedDerivatives implementations are untested
+
+// If rendering directly to the swapchain, dFdx/dFdy operations may be incorrect due to the swapchain
+// transform. Use these helper functions to transform the dFdx or dFdy accordingly.
+
+float2 transformedDerivatives(in uint swapchainTransform, in float2 ddxDdy)
+{
+    #define OUTPUT_TYPE float2
+    #include "nbl/builtin/hlsl/utils/surface_transform_transformedDerivatives.hlsl"
+    #undef OUTPUT_TYPE
+}
+float2x2 transformedDerivatives(in uint swapchainTransform, in float2x2 ddxDdy)
+{
+    #define OUTPUT_TYPE float2x2
+    #include "nbl/builtin/hlsl/utils/surface_transform_transformedDerivatives.hlsl"
+    #undef OUTPUT_TYPE
+}
+float2x3 transformedDerivatives(in uint swapchainTransform, in float2x3 ddxDdy)
+{
+    #define OUTPUT_TYPE float2x3
+    #include "nbl/builtin/hlsl/utils/surface_transform_transformedDerivatives.hlsl"
+    #undef OUTPUT_TYPE
+}
+float2x4 transformedDerivatives(in uint swapchainTransform, in float2x4 ddxDdy)
+{
+    #define OUTPUT_TYPE float2x4
+    #include "nbl/builtin/hlsl/utils/surface_transform_transformedDerivatives.hlsl"
+    #undef OUTPUT_TYPE
+}
```

Comment:
you'll no longer need "nbl/builtin/hlsl/utils/surface_transform_transformedDerivatives.hlsl" to exist, this is doable with a template